template.py: Fix error thrown if default passed to compile_template i…

github-abcde opened the pull request:
### What does this PR do?
It reduces errors logged when using a renderer pipeline with more than one renderer where this renderer does not produce data of type cStringIO (and therefor does not have a .seek method).
It renders data as-is when the data has a shebang, but none of the renderers are recognized or all are blacklisted.
It causes the renderer to render the data as-is when the default renderer supplied is None.

### What issues does this PR fix or reference?
None

### Previous Behavior
Files rendered with a shebang where all mentioned renderers are blacklisted are rendered with the default renderer pipeline instead.
Error logged ([ERROR   ] error while compiling template '/path/to/file': 'str' object has no attribute 'seek') when using multiple renderers in a pipeline where the rendered data is not of type cStringIO.

### New Behavior
Files rendered with a shebang where all mentioned renderers are blacklisted are rendered as-is
No more "no attribute 'seek'"-errors.

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

Commit:
template.py: Fix error thrown if default passed to compile_template is None. Altered functionality to render the file as-is when no renderers in the shebang are recognized or survive blacklist-filtering. Removed error while compiling template that was the cause of a call to .seek when the datatype does not have this method.

cachedout commented on salt/template.py:
```diff
@@ -88,8 +88,9 @@ def compile_template(template,
                 input_data.seek(0)
-            except Exception as exp:
-                log.error('error while compiling template \'{0}\': {1}'.format(template, exp))
+            except AttributeError:
```
I can see the case for catching and ignoring an `AttributeError` but I don't follow why we would also remove the generalized `Exception` handling. Could you explain why that's necessary here?

github-abcde replied:
The only possible other exception that could be thrown here is an IOError, in case the object is seekable, but it fails for some reason. But since this is an in-memory-string handled as a file object, that is (imo) very unlikely to happen. This lead me to conclude that a generic exception is out of place here.
I do not quite understand the comment at https://github.com/saltstack/salt/commit/99024ba1c86dba21550c28f4998aaa0cd0f0169a which could shed some light on why a catch-all is here in the first place (even though it was silenced).

cachedout replied:
@basepi Do you remember by chance?

basepi replied:
I do not remember what error I was getting that caused me to add the generalized exception handling. I was probably just being lazy about researching which exceptions could be thrown. It might have been an AttributeError, I just can't remember.

martin-paulus replied:
@cachedout Since the reasoning for the general exception handling is lost, do you still require a code change to accept this PR?

cachedout replied:
@martin-paulus I think it's fine. I just wanted to research this as best we could first. :]

Commit:
template.py: Changed approach to seeking and logging. Now only log and call .seek if input_data (and ret when logging) is of type cStringIO.InputType.

github-abcde commented:
Ok, how about this? Instead of try-catching errors when the datatype is not cStringIO, let's only seek (and log) when the datatype *is* cStringIO.

Commit:
template.py: Included StringIO.StringIO next to cStringIO.InputType.
utils/templates.py: Added .encode(SLS_ENCODING) to avoid templating string being encoded with ASCII codec, which then borks on UTF-encoded content.

cachedout commented:
It does look like we're seeing failures in the template tests now, unfortunately. https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/7804/

github-abcde commented:
This one: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/7804/testReport/junit/unit.templates.jinja_test/TestGetTemplate/test_non_ascii/
is the reason I edited salt/utils/templates.py and added the .encode. But since the test uses a file that's written halfway in the templating process, it doesn't use the same result as is being returned.

Commit:
utils/templates.py:wrap_tmpl_func: Moved manual encode from the return-value to the output of render_str. Removed SLS_ENCODER(output) because for some reason it ends up using the ascii-codec (which it shouldnt). Unit tests pass again.

github-abcde commented:
Latest changes fixes the test failures that happened. Except the test which fails because __opts__['grains'] doesn't seem to exist, but that seems to be an error outside the scope of the changes introduced in this PR.

cachedout commented:
Go Go Jenkins!
